src/core/qgsmaplayerregistry.cpp

```diff
@@ -203,14 +220,17 @@ void QgsMapLayerRegistry::onMapLayerDeleted( QObject* obj )
 
 QMap<QString, QgsMapLayer*> QgsMapLayerRegistry::mapLayers()
 {
-  return mMapLayers;
+  QReadLocker lock( &mRegistryLock );
+
+  QMap<QString, QgsMapLayer*> mapLayers = mMapLayers;
+  mapLayers.detach();
+  return mapLayers;
```

Comment:
Is detach thread safe?

I.e. what happens if thread A removes a layer and detaches while thread B is iterating over the container, is there no chance for a race condition?